src/app/use-cases/validate-document.ts

```diff
@@ -0,0 +1,75 @@
+import * as TE from "fp-ts/TaskEither";
+import * as O from "fp-ts/Option";
+import * as E from "fp-ts/Either";
+import { identity, pipe } from "fp-ts/function";
+
+import { findIndex, modifyAt } from "fp-ts/Array";
+import { validate } from "@pagopa/handler-kit/lib/validation";
+import {
+  GetSignatureRequest,
+  SignatureRequest,
+  UpsertSignatureRequest,
+} from "../../signature-request/signature-request";
+
+import { EntityNotFoundError } from "../../error";
+
+import {
+  Document,
+  DocumentList,
+  IsDocumentUploaded,
+} from "../../signature-request/document";
+import { GetDocumentPayload } from "./get-document";
+
+export const addUrlToDocument =
+  (documentId: Document["id"], url: string) => (request: SignatureRequest) =>
+    pipe(
+      request.documents,
+      findIndex((document) => document.id === documentId),
+      O.chain((index) =>
+        pipe(
+          request.documents,
+          modifyAt(index, (document) => ({
+            ...document,
+            url,
+          }))
+        )
+      ),
+      E.fromOption(() => new EntityNotFoundError("Document not found")),
+      E.chainW(validate(DocumentList, "Unable to validate document list"))
+    );
+
+export type ValidateDocumentPayload = GetDocumentPayload & {
+  documentUrl: string;
+};
+
+export const makeValidateDocument =
+  (
+    getSignatureRequest: GetSignatureRequest,
+    upsertSignatureRequest: UpsertSignatureRequest,
+    isDocumentUploaded: IsDocumentUploaded
+  ) =>
+  (payload: ValidateDocumentPayload) =>
+    pipe(
+      isDocumentUploaded(payload.documentId),
+      TE.filterOrElse(
+        identity,
+        () => new Error("Unable to find the uploaded document")
+      ),
+      TE.chain(() =>
+        getSignatureRequest(payload.signatureRequestId, payload.subscriptionId)
+      ),
+      TE.chainW(
+        TE.fromOption(
+          () => new EntityNotFoundError("Signature request not found")
```

Comment:
what do you mean by `grouping`?

Reply:
for example, create a file containing all errors as global variables. In this way, in case of modification of the error message, it is sufficient to modify it in one point only